File: settlement/room/military/supply/ROOM_SUPPLY.java
package settlement.room.military.supply;

import java.io.IOException;

import game.GAME;
import init.resources.RBIT.RBITImp;
import init.resources.RESOURCE;
import init.resources.RESOURCES;
import settlement.path.finders.SFinderRoomService;
import settlement.room.main.Room;
import settlement.room.main.RoomBlueprintIns;
import settlement.room.main.category.RoomCategorySub;
import settlement.room.main.furnisher.Furnisher;
import settlement.room.main.job.ROOM_EMPLOY_AUTO;
import settlement.room.main.job.ROOM_RADIUS;
import settlement.room.main.util.RoomInitData;
import snake2d.util.file.FileGetter;
import snake2d.util.file.FilePutter;
import snake2d.util.sets.LISTE;
import view.sett.ui.room.UIRoomModule;

public final class ROOM_SUPPLY extends RoomBlueprintIns<SupplyInstance> implements ROOM_RADIUS, ROOM_EMPLOY_AUTO{

	public static final int STORAGE = 80;
	final Constructor constructor;
	final Crate crate = new Crate(this);
	final Cache cache = new Cache(this);
	public final SupplyTally tally = new SupplyTally(this);
	final RESOURCE liveStock;
	
	public ROOM_SUPPLY(RoomInitData init, RoomCategorySub cat) throws IOException {
		super(0, init, "_MILITARY_SUPPLY", cat);
		constructor = new Constructor(this, init);
		liveStock = RESOURCES.map().read("LIVESTOCK", init.data());
		

	}

	@Override
	protected void update(double ds) {
		
		
	}
	
	@Override
	public Furnisher constructor() {
		return constructor;
	}

	@Override
	public SFinderRoomService service(int tx, int ty) {
		return null;
	}

	@Override
	protected void saveP(FilePutter saveFile){
		cache.save(saveFile);
	}
	
	@Override
	protected void loadP(FileGetter saveFile) throws IOException{
		cache.clear();
		tally.clear();
		for (SupplyInstance ins : all()) {
			tally.init(ins);
		}
		cache.load(saveFile);
	}
	
	@Override
	protected void clearP() {
		cache.clear();
		tally.clear();
		cache.clear();
	}
	
	@Override
	public boolean degrades() {
		return false;
	}
	
	@Override
	public void appendView(LISTE<UIRoomModule> mm) {
		mm.add(new Gui(this).make());
	}

	@Override
	public boolean autoEmploy(Room r) {
		return ((SupplyInstance)r).auto;
	}

	@Override
	public void autoEmploy(Room r, boolean b) {
		((SupplyInstance)r).auto = b;
	}
	
//	void go(Crate cr) {
//
//		int tx = cr.crate.x();
//		int ty = cr.crate.y();
//		
//		int am = cr.reserveGo();
//		if (am <= 0) {
//			return;
//		}
//		DIR dir = DIR.ORTHO.get(SETT.ROOMS().fData.item.get(tx, ty).rotation);
//		byte ran = (byte) SETT.tileRan(tx, ty);
//		if (!SETT.HALFENTS().transports.military(tx+dir.x()*2, ty+dir.y()*2, ran, cr.crate.resource(), am, dir)) {
//			cr = crate.get(tx, ty);
//			cr.goCancel(am);
//		}
//		cr = crate.get(tx, ty);
//		cr.go(am);
//		
//	}
	
//	public int goReserve(int mx, int my, ShortCoo coo) {
//		SupplyInstance ins = get(mx, my);
//		if (ins == null)
//			return 0;
//		if (ins.goCount >= ins.jobs.size())
//			ins.goCount = 0;
//		COORDINATE c = ins.jobs.get(ins.goCount);
//		coo.set(c);
//		ins.goCount++;
//		Crate cr = crate.get(c.x(), c.y());
//		int am = cr.reserveGo();
//		return am;
//	}
//	
//	public RESOURCE goCrate(COORDINATE c, int am){
//		
//		Crate cr = crate.get(c.x(), c.y());
//		if (cr == null)
//			return null;
//		
//		return cr.go(am);
//	}
//	
//	public void goCancel(COORDINATE c, int am) {
//		Crate cr = crate.get(c.x(), c.y());
//		if (cr == null)
//			return;
//		cr.goCancel(am);
//	}
	
	private int upI = -1;
	private RBITImp hh = new RBITImp();
	
	public boolean has(RESOURCE res) {
		if (upI == GAME.updateI()) {
			return hh.has(res);
		}
		upI = GAME.updateI();
		for (int i = 0; i < instancesSize(); i++) {
			hh.or(getInstance(i).allowed());
		}
		return hh.has(res);
	}
	

}

File: LegacySource/settlement/room/military/supply/ROOM_SUPPLY.cs
using System;
using System.Collections.Generic;
using System.IO;
using game;
using init.resources;
using settlement.path.finders;
using settlement.room.main;
using settlement.room.main.category;
using settlement.room.main.furnisher;
using settlement.room.main.job;
using settlement.room.main.util;
using snake2d.util.file;
using snake2d.util.sets;
using view.sett.ui.room;

namespace settlement.room.military.supply
{
    public sealed class ROOM_SUPPLY : RoomBlueprintIns<SupplyInstance>, ROOM_RADIUS, ROOM_EMPLOY_AUTO
    {
        public const int STORAGE = 80;
        private readonly Constructor constructor;
        private readonly Crate crate;
        private readonly Cache cache;
        public readonly SupplyTally tally;
        private readonly RESOURCE liveStock;

        public ROOM_SUPPLY(RoomInitData init, RoomCategorySub cat) : base(0, init, "_MILITARY_SUPPLY", cat)
        {
            constructor = new Constructor(this, init);
            liveStock = RESOURCES.map().read("LIVESTOCK", init.data());
        }

        protected override void update(double ds)
        {
            
        }

        public override Furnisher constructor()
        {
            return constructor;
        }

        public override SFinderRoomService service(int tx, int ty)
        {
            return null;
        }

        protected override void saveP(FilePutter saveFile)
        {
            cache.save(saveFile);
        }

        protected override void loadP(FileGetter saveFile)
        {
            cache.clear();
            tally.clear();
            foreach (var ins in all())
            {
                tally.init(ins);
            }
            cache.load(saveFile);
        }

        protected override void clearP()
        {
            cache.clear();
            tally.clear();
            cache.clear();
        }

        public override bool degrades()
        {
            return false;
        }

        public override void appendView(LISTE<UIRoomModule> mm)
        {
            mm.add(new Gui(this).make());
        }

        public override bool autoEmploy(Room r)
        {
            return ((SupplyInstance)r).auto;
        }

        public override void autoEmploy(Room r, bool b)
        {
            ((SupplyInstance)r).auto = b;
        }

        private int upI = -1;
        private RBITImp hh = new RBITImp();

        public bool has(RESOURCE res)
        {
            if (upI == GAME.updateI())
            {
                return hh.has(res);
            }
            upI = GAME.updateI();
            for (int i = 0; i < instancesSize(); i++)
            {
                hh.or(getInstance(i).allowed());
            }
            return hh.has(res);
        }
    }
}